File: #54_spiralOrder.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []
        else:
            cols = len(matrix[0])
            rows = len(matrix)
            res = []
            firstRow, firstCol = 0, 0
            while rows>0 and cols>0:
                lastRow = firstRow+rows-1
                lastCol = firstCol+cols-1

                for i in range(cols):
                    res.append(matrix[firstRow][firstCol+i])
                for i in range(1, rows):
                    res.append(matrix[firstRow+i][lastCol])
                for i in range(1,cols):
                    if firstRow!=lastRow:
                        res.append(matrix[lastRow][lastCol-i])
                for i in range(1,rows-1):
                    if firstCol !=lastCol:
                        res.append(matrix[lastRow-i][firstCol])

                firstRow +=1
                firstCol +=1
                rows -=2
                cols -=2

            return res

File: Java/#54_spiralOrder.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<Integer>();
        int rowStart=0;
        int rowEnd = matrix.length-1;
        int colStart = 0;
        int colEnd=matrix[0].length-1;
        while(rowStart<=rowEnd && colStart<=colEnd){
            for(int c=colStart;c<=colEnd;c++){
                res.add(matrix[rowStart][c]);
            }
            for(int r=rowStart+1;r<=rowEnd;r++){
                res.add(matrix[r][colEnd]);
            }
            if(rowStart<rowEnd && colStart<colEnd){
                for(int c=colEnd-1;c>colStart;c--){
                    res.add(matrix[rowEnd][c]);
                }
                for(int r=rowEnd;r>rowStart;r--){
                   res.add(matrix[r][colStart]);
                }
            }
            rowStart++;
            rowEnd--;
            colStart++;
            colEnd--;
        }
        return res;
    }
}
